## Exercise L3 - 1: Diagnose Dataset Level and Select Last Encounter

### Instructions
- Given the dataset, convert the dataset to a longitudinal level but select only the last encounter for each patient.
- Assume that that the order of encounter IDs is indicative of the time for encounter. In other words a lower number encounter will come before a higher numbered encounter.

In [1]:
import pandas as pd
import numpy as np

In [2]:
ehr_level_dataset_path = "./data/ehr_level_exercise_dataset.csv"

### Level of Dataset
What level is the dataset at? Is at the line or encounter level? 

### Solution

In [3]:
ehr_level_df = pd.read_csv(ehr_level_dataset_path)
ehr_level_df.head()

ENCOUNTER_ID                     PATIENT_ID  \
0     udacity_health_encounter_id_1  udacity_health_patient_id_186   
1    udacity_health_encounter_id_10  udacity_health_patient_id_188   
2   udacity_health_encounter_id_100  udacity_health_patient_id_585   
3  udacity_health_encounter_id_1000  udacity_health_patient_id_525   
4  udacity_health_encounter_id_1001  udacity_health_patient_id_950   

  PRINCIPAL_DIAGNOSIS_CODE                                     PROCEDURE_CODE  \
0            dx_code_15406  ['procedure_code_58552', 'procedure_code_39776...   
1            dx_code_74047                                                 []   
2            dx_code_71465                                                 []   
3            dx_code_61569                                                 []   
4            dx_code_90172                           ['procedure_code_30555']   

                                     MEDICATION_CODE  \
0  ['medication_code_2350', 'medication_code_8630...   
1  ['medication_code_7789', 'medication_code_3560...   
2  ['medication_code_7982', 'medication_code_2452...   
3                           ['medication_code_4036']   
4   ['medication_code_6755', 'medication_code_5045']   

                                            LAB_CODE  \
0  ['lab_code_8835', 'lab_code_9859', 'lab_code_9...   
1                                                 []   
2  ['lab_code_4198', 'lab_code_6603', 'lab_code_3...   
3                                                 []   
4                                  ['lab_code_9112']   

                                               LABEL  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
1                                       [1, 1, 1, 1]  
2                              [0, 0, 0, 0, 0, 0, 0]  
3                                                [0]  
4                                       [0, 0, 0, 0]

**Tests**
- Line: Total number of rows > Number of Unique Encounters
- Encounter level: Total Number of Rows = Number of Unique Encounters


In [4]:
# Line Test
try:
    assert len(ehr_level_df) > ehr_level_df['ENCOUNTER_ID'].nunique() 
    print("Dataset could be at the line level")
except:
    print("Dataset is not at the line level")

Dataset is not at the line level


In [5]:
# Encounter Test
try:
    assert len(ehr_level_df) == ehr_level_df['ENCOUNTER_ID'].nunique()
    print("Dataset could be at the encounter level")
except:
    print("Dataset is not at the encounter level")

Dataset could be at the encounter level


**Answer:** Dataset is at the encounter level and you can probably guess by seeing the arrays for the code sets but we did a few simple tests to confirm.

### Select Last Encounter for each Patient

So in many cases you may only want a snapshot of a patient's history for your modeling objective. In some cases it might be important to see the changes over time but in other cases you only want the most recent case or depending on the model the first case could also be used. Really important to know how the context for how the model will be deployed in production and the time state you will be getting data.

In [6]:
# select last encounter for each patient
#convert encounter id column to a numerical value
def convert_encounter_id_to_number(df, encounter_id):
    df["ENCOUNTER_ID_NUMBER"] = df[encounter_id].str.replace('udacity_health_encounter_id_', '').astype(int)
    return df

def select_last_encounter(df, patient_id, encounter_id):
    df = df.sort_values(encounter_id)
    last_encounter_values = df.groupby(patient_id)[encounter_id].tail(1).values
    return df[df[encounter_id].isin(last_encounter_values)] 

In [7]:
ehr_encounter_number_df = convert_encounter_id_to_number(ehr_level_df, "ENCOUNTER_ID")
last_encounter_df = select_last_encounter(ehr_encounter_number_df, "PATIENT_ID", "ENCOUNTER_ID_NUMBER" )

In [8]:
#take subset of output
test_last_encounter_df = last_encounter_df[['ENCOUNTER_ID', 'ENCOUNTER_ID_NUMBER', 'PATIENT_ID']]

### Test cases 
- PATIENT_IDS - udacity_health_patient_id_309, udacity_health_patient_id_418, udacity_health_patient_id_908

In [9]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_309']

ENCOUNTER_ID                     PATIENT_ID  \
94    udacity_health_encounter_id_1091  udacity_health_patient_id_309   
450   udacity_health_encounter_id_1436  udacity_health_patient_id_309   
605   udacity_health_encounter_id_1582  udacity_health_patient_id_309   
665   udacity_health_encounter_id_1643  udacity_health_patient_id_309   
684   udacity_health_encounter_id_1662  udacity_health_patient_id_309   
867   udacity_health_encounter_id_1837  udacity_health_patient_id_309   
1024  udacity_health_encounter_id_1985  udacity_health_patient_id_309   
1042  udacity_health_encounter_id_2000  udacity_health_patient_id_309   
1196  udacity_health_encounter_id_2142  udacity_health_patient_id_309   
1289  udacity_health_encounter_id_2229  udacity_health_patient_id_309   
1340  udacity_health_encounter_id_2281  udacity_health_patient_id_309   
1596  udacity_health_encounter_id_2531  udacity_health_patient_id_309   
1702  udacity_health_encounter_id_2631  udacity_health_patient_id_309   
2069  udacity_health_encounter_id_2989  udacity_health_patient_id_309   
2291  udacity_health_encounter_id_3209  udacity_health_patient_id_309   
2507  udacity_health_encounter_id_3416  udacity_health_patient_id_309   
2696   udacity_health_encounter_id_360  udacity_health_patient_id_309   
2890  udacity_health_encounter_id_3792  udacity_health_patient_id_309   
2960  udacity_health_encounter_id_3859  udacity_health_patient_id_309   
3064  udacity_health_encounter_id_3961  udacity_health_patient_id_309   
3167  udacity_health_encounter_id_4058  udacity_health_patient_id_309   
3280  udacity_health_encounter_id_4168  udacity_health_patient_id_309   
3353  udacity_health_encounter_id_4239  udacity_health_patient_id_309   
3561  udacity_health_encounter_id_4439  udacity_health_patient_id_309   
3585  udacity_health_encounter_id_4462  udacity_health_patient_id_309   
3718  udacity_health_encounter_id_4589  udacity_health_patient_id_309   
3752  udacity_health_encounter_id_4619  udacity_health_patient_id_309   
3916   udacity_health_encounter_id_478  udacity_health_patient_id_309   
4552  udacity_health_encounter_id_5398  udacity_health_patient_id_309   
4687   udacity_health_encounter_id_553  udacity_health_patient_id_309   
4771  udacity_health_encounter_id_5611  udacity_health_patient_id_309   
4785  udacity_health_encounter_id_5624  udacity_health_patient_id_309   
4796  udacity_health_encounter_id_5635  udacity_health_patient_id_309   
4862  udacity_health_encounter_id_5701  udacity_health_patient_id_309   
4881  udacity_health_encounter_id_5721  udacity_health_patient_id_309   
5132   udacity_health_encounter_id_597  udacity_health_patient_id_309   
5306  udacity_health_encounter_id_6139  udacity_health_patient_id_309   
5385   udacity_health_encounter_id_622  udacity_health_patient_id_309   
5472  udacity_health_encounter_id_6305  udacity_health_patient_id_309   
5564  udacity_health_encounter_id_6394  udacity_health_patient_id_309   
5669  udacity_health_encounter_id_6497  udacity_health_patient_id_309   
5684  udacity_health_encounter_id_6512  udacity_health_patient_id_309   
5829  udacity_health_encounter_id_6662  udacity_health_patient_id_309   
5846   udacity_health_encounter_id_668  udacity_health_patient_id_309   
5946  udacity_health_encounter_id_6777  udacity_health_patient_id_309   
6508  udacity_health_encounter_id_7327  udacity_health_patient_id_309   
6892  udacity_health_encounter_id_7696  udacity_health_patient_id_309   
6969  udacity_health_encounter_id_7772  udacity_health_patient_id_309   
7217   udacity_health_encounter_id_995  udacity_health_patient_id_309   

     PRINCIPAL_DIAGNOSIS_CODE  \
94              dx_code_62502   
450             dx_code_37825   
605             dx_code_79663   
665             dx_code_60773   
684             dx_code_29272   
867             dx_code_56368   
1024            dx_code_38581   
1042            dx_code_91230   
1196            dx_code_58527   
1289            dx_code_46826   
1340        

For patient id 309, the selected encounter should be 7772.

In [10]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_309']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
6969  udacity_health_encounter_id_7772                 7772   

                         PATIENT_ID  
6969  udacity_health_patient_id_309

In [11]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_418']

ENCOUNTER_ID                     PATIENT_ID  \
263   udacity_health_encounter_id_1258  udacity_health_patient_id_418   
1230  udacity_health_encounter_id_2173  udacity_health_patient_id_418   
2448  udacity_health_encounter_id_3362  udacity_health_patient_id_418   

     PRINCIPAL_DIAGNOSIS_CODE  \
263             dx_code_87320   
1230            dx_code_72104   
2448            dx_code_77979   

                                         PROCEDURE_CODE  \
263                                                  []   
1230  ['procedure_code_37393', 'procedure_code_48757...   
2448                           ['procedure_code_64397']   

                                        MEDICATION_CODE  \
263                            ['medication_code_7475']   
1230  ['medication_code_3809', 'medication_code_9141...   
2448   ['medication_code_2807', 'medication_code_6979']   

                                               LAB_CODE  \
263                                                  []   
1230  ['lab_code_2839', 'lab_code_2665', 'lab_code_3...   
2448                                                 []   

                                                  LABEL  ENCOUNTER_ID_NUMBER  
263                                                 [0]                 1258  
1230  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...                 2173  
2448                                          [0, 0, 0]                 3362

For patient id 418, the selected encounter should be 3362.

In [12]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_418']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
2448  udacity_health_encounter_id_3362                 3362   

                         PATIENT_ID  
2448  udacity_health_patient_id_418

In [13]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_908']

ENCOUNTER_ID                     PATIENT_ID  \
1678  udacity_health_encounter_id_2608  udacity_health_patient_id_908   
2796  udacity_health_encounter_id_3698  udacity_health_patient_id_908   
5301  udacity_health_encounter_id_6132  udacity_health_patient_id_908   
7131   udacity_health_encounter_id_910  udacity_health_patient_id_908   

     PRINCIPAL_DIAGNOSIS_CODE  \
1678            dx_code_52877   
2796            dx_code_52688   
5301            dx_code_36066   
7131            dx_code_98539   

                                         PROCEDURE_CODE  \
1678                                                 []   
2796                           ['procedure_code_60912']   
5301  ['procedure_code_5724', 'procedure_code_55191'...   
7131  ['procedure_code_8188', 'procedure_code_59872'...   

                                        MEDICATION_CODE  \
1678                           ['medication_code_6189']   
2796                                                 []   
5301  ['medication_code_9532', 'medication_code_3398...   
7131  ['medication_code_5112', 'medication_code_3537...   

                                               LAB_CODE  \
1678                                                 []   
2796                 ['lab_code_4691', 'lab_code_9380']   
5301  ['lab_code_9732', 'lab_code_2808', 'lab_code_3...   
7131                                                 []   

                                           LABEL  ENCOUNTER_ID_NUMBER  
1678                                         [0]                 2608  
2796                                   [0, 0, 0]                 3698  
5301  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]                 6132  
7131                       [0, 0, 0, 0, 0, 0, 0]                  910

For patient id 908, the selected encounter should be 6132.

In [14]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_908']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
5301  udacity_health_encounter_id_6132                 6132   

                         PATIENT_ID  
5301  udacity_health_patient_id_908

## Exercise  L3 - 2: Dataset Splitting

### Instructions
- Split the provided dataset into a train and test split but be sure not to mix patient encounter records across the two partitions
- Be sure to run the following three tests
    - Patient data in only one partition
    - Total unique number of patients across all partitions = total number unique patients in original full dataset
    - Total number of rows original dataset = sum of rows across splits

In [15]:
splitting_exercise_dataset_path = "./data/SYNTHETIC_EHR_DATASET.csv"

### Solution

This is largely a review of two parts in this lesson and you can use most of the same code for each step. The key is to identify the level of the dataset and then to convert it to the encounter level before you do your splits. Then perform the splitting and run the tests.

#### Convert to Encounter Level

In [16]:
# convert to encounter and then split but make sure 
ehr_pre_split_df = pd.read_csv(splitting_exercise_dataset_path)
grouping_field_list = ['ENCOUNTER_ID', 'PATIENT_ID', 'PRINCIPAL_DIAGNOSIS_CODE']
non_grouped_field_list = [c for c in ehr_pre_split_df.columns if c not in grouping_field_list]
ehr_encounter_df = ehr_pre_split_df.groupby(grouping_field_list)[non_grouped_field_list].agg(lambda x: 
                                                        list([y for y in x if y is not np.nan ] ) ).reset_index()

#### Split at Patient Level

In [17]:
PATIENT_ID_FIELD = 'PATIENT_ID'
TEST_PERCENTAGE = 0.2

In [18]:
def split_dataset_patient_level(df, key, test_percentage=0.2):
    df = df.iloc[np.random.permutation(len(df))]
    unique_values = df[key].unique()
    total_values = len(unique_values)
    sample_size = round(total_values * (1 - test_percentage ))
    train = df[df[key].isin(unique_values[:sample_size])].reset_index(drop=True)
    test = df[df[key].isin(unique_values[sample_size:])].reset_index(drop=True)
    return train, test

In [19]:
train_df, test_df = split_dataset_patient_level(ehr_encounter_df, PATIENT_ID_FIELD, TEST_PERCENTAGE)

In [20]:
assert len(set(train_df[PATIENT_ID_FIELD].unique()).intersection(set(test_df[PATIENT_ID_FIELD].unique()))) == 0
print("Test passed for patient data in only one partition")

Test passed for patient data in only one partition


In [21]:
assert (train_df[PATIENT_ID_FIELD].nunique()  + test_df[PATIENT_ID_FIELD].nunique()) == ehr_encounter_df[PATIENT_ID_FIELD].nunique()
print("Test passed for number of unique patients being equal!")

Test passed for number of unique patients being equal!


In [22]:
assert len(train_df)  + len(test_df) == len(ehr_encounter_df)
print("Test passed for number of total rows equal!")

Test passed for number of total rows equal!


#### Optional
- Check label distribution and use scikitlearn - https://scikit-learn.org/stable/auto_examples/model_selection/plot_cv_indices.html#sphx-glr-auto-examples-model-selection-plot-cv-indices-py

## Exercise  L3 - 3: Build Bucketed Numeric Feature with TF

### Instructions
- Given the Swiss heart disease dataset that we worked with earlier, build a bucketed numeric feature from the age feature. 
- For this exercise, use the Tensorflow csv function for loading the dataset directly into a TF tensor -https://www.tensorflow.org/api_docs/python/tf/data/experimental/make_csv_dataset. This approach will be useful for when you have much larger datasets and also allows you to bypass loading the dataset in Pandas.
- More information on the Tensorflow bucketized feature can be found here https://www.tensorflow.org/api_docs/python/tf/feature_column/bucketized_column. Bucketed features take as input the  numeric feature that we covered in the lesson. For the numeric feature, you do not need to normalize it like we did in the lesson. 

In [23]:
import tensorflow as tf
swiss_dataset_path = "./data/lesson_exercise_swiss_dataset.csv"
BATCH_SIZE =128
PREDICTOR_FIELD = 'num_label'

### Solution

In [24]:
# ETL with TF dataset make csv function
swiss_tf_dataset = tf.data.experimental.make_csv_dataset( swiss_dataset_path, batch_size=BATCH_SIZE, 
                                                         num_epochs=1, label_name=PREDICTOR_FIELD, header=True)

In [25]:
swiss_dataset_batch = next(iter(swiss_tf_dataset))[0]
swiss_dataset_batch

OrderedDict([('age',
              <tf.Tensor: shape=(123,), dtype=int32, numpy=
              array([63, 72, 60, 38, 59, 67, 62, 52, 54, 73, 55, 50, 42, 35, 61, 60, 46,
                     41, 60, 60, 62, 45, 53, 59, 61, 62, 61, 43, 69, 38, 56, 60, 62, 74,
                     52, 54, 51, 61, 66, 56, 53, 50, 38, 61, 70, 57, 58, 51, 53, 32, 51,
                     60, 48, 59, 68, 56, 65, 59, 40, 63, 55, 69, 63, 53, 50, 46, 62, 51,
                     68, 53, 56, 55, 51, 58, 57, 64, 47, 64, 61, 43, 34, 47, 70, 56, 57,
                     51, 47, 65, 38, 66, 61, 57, 50, 56, 47, 63, 56, 62, 38, 62, 61, 43,
                     58, 53, 59, 61, 63, 51, 36, 57, 65, 52, 64, 38, 53, 65, 56, 55, 52,
                     53, 42, 57, 54], dtype=int32)>)])

In [26]:
# create TF numeric feature
tf_numeric_age_feature = tf.feature_column.numeric_column(key='age', default_value=0, dtype=tf.float64)

In [27]:
#boundaries for the different age buckets
b_list = [ 0, 18, 25, 40, 55, 65, 80, 100]
#create TF bucket feature from numeric feature
tf_bucket_age_feature = tf.feature_column.bucketized_column(source_column=tf_numeric_age_feature, boundaries= b_list)

In [28]:
def demo(feature_column, example_batch):
    feature_layer = tf.keras.layers.DenseFeatures(feature_column)
    print(feature_layer(example_batch))
    print("\nExample of one transformed row:")
    print(feature_layer(example_batch).numpy()[0])

In [29]:
print("Example bucket field:\n{}\n".format(tf_bucket_age_feature))
demo(tf_bucket_age_feature, swiss_dataset_batch)

Example bucket field:
BucketizedColumn(source_column=NumericColumn(key='age', shape=(1,), default_value=(0,), dtype=tf.float64, normalizer_fn=None), boundaries=(0, 18, 25, 40, 55, 65, 80, 100))

tf.Tensor(
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]], shape=(123, 9), dtype=float32)

Example of one transformed row:
[0. 0. 0. 0. 0. 1. 0. 0. 0.]


## Exercise  L3 - 4:  Build Embedding Categorical Feature with TF

### Instructions
- Build a 10 dimension embedding feature for the PRINCIPAL_DIAGNOSIS_CODE field
- Here is the link to the Tensorflow Embedding column documentation -https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column
- Some functions provided below to assist

In [30]:
ehr_line_df = pd.read_csv("./data/SYNTHETIC_EHR_DATASET.csv")
cat_example_df = ehr_line_df[['ENCOUNTER_ID', 'PRINCIPAL_DIAGNOSIS_CODE', 'LABEL']]

In [31]:
#adapted from https://www.tensorflow.org/tutorials/structured_data/feature_columns
def df_to_dataset(df, predictor,  batch_size=32):
    df = df.copy()
    labels = df.pop(predictor)
    ds = tf.data.Dataset.from_tensor_slices((dict(df), labels))
    ds = ds.shuffle(buffer_size=len(df))
    ds = ds.batch(batch_size)
    return ds

In [32]:
BATCH_SIZE = 64
PREDICTOR_FIELD = 'LABEL'
categorical_tf_ds = df_to_dataset(cat_example_df, PREDICTOR_FIELD, batch_size=BATCH_SIZE)

In [33]:
# build vocab for categorical features
def write_vocabulary_file(vocab_list, field_name, default_value, vocab_dir='./vocab/'):
    output_file_path = os.path.join(vocab_dir, str(field_name) + "_vocab.txt")
    # put default value in first row as TF requires
    vocab_list = np.insert(vocab_list, 0, default_value, axis=0) 
    df = pd.DataFrame(vocab_list).to_csv(output_file_path, index=None, header=None)
    return output_file_path

def build_vocab_files(df, categorical_column_list, default_value='00'):
    vocab_files_list = []
    for c in categorical_column_list:
        v_file = write_vocabulary_file(df[c].unique(), c, default_value)
        vocab_files_list.append(v_file)
    return vocab_files_list

### Solution

In [34]:
import os
# add logic to add if not exist
#os.mkdir("./vocab/")

In [35]:
categorical_field_list = ["PRINCIPAL_DIAGNOSIS_CODE"]
vocab_files_list = build_vocab_files(cat_example_df, categorical_field_list)
vocab_files_list[0]

'./vocab/PRINCIPAL_DIAGNOSIS_CODE_vocab.txt'

In [36]:
principal_diagnosis_vocab = tf.feature_column.categorical_column_with_vocabulary_file(
            key="PRINCIPAL_DIAGNOSIS_CODE", vocabulary_file = vocab_files_list[0], num_oov_buckets=1)

INFO:tensorflow:vocabulary_size = 6753 in PRINCIPAL_DIAGNOSIS_CODE is inferred from the number of elements in the vocabulary_file ./vocab/PRINCIPAL_DIAGNOSIS_CODE_vocab.txt.


In [37]:
dims = 10
cat_embedded = tf.feature_column.embedding_column(principal_diagnosis_vocab, dimension=dims)

In [38]:
categorical_tf_ds_batch = next(iter(categorical_tf_ds))[0]

In [39]:
demo(cat_embedded, categorical_tf_ds_batch)

tf.Tensor(
[[-0.27406758  0.4812879  -0.4463131  -0.01966097 -0.2481951   0.15411611
   0.40982312 -0.23224503  0.21873532  0.00971104]
 [ 0.2426273   0.37905917  0.02748485  0.34419003  0.17442113  0.12028006
   0.26426268 -0.23395902 -0.1674242   0.26317292]
 [ 0.08327362  0.09212287  0.1035822   0.04222572 -0.10462294  0.24624164
   0.22006914  0.06731185 -0.3662947  -0.4294941 ]
 [ 0.5373018  -0.47315276  0.3481966   0.0062991   0.4224541  -0.09058967
   0.00871684  0.264543   -0.01122587 -0.03040619]
 [-0.5202304   0.2002285   0.03574209 -0.38446787  0.31543237 -0.07109011
   0.40516222  0.22675845  0.0548689  -0.58886874]
 [-0.09681373  0.21142933  0.27343258 -0.07801515  0.11138771 -0.34336254
   0.30648053 -0.28212842 -0.1954338  -0.2991298 ]
 [ 0.36081544 -0.41307458  0.16407725 -0.38049543 -0.50218034  0.38408303
  -0.30005577  0.10407525 -0.61177635 -0.06910674]
 [ 0.10818396  0.04000156  0.2088314  -0.58167964 -0.25659883 -0.26644713
  -0.21687894  0.35785598 -0.17677933 -0